contrib/guix/libexec/build.sh

```diff
@@ -452,11 +452,7 @@ mv --no-target-directory "$OUTDIR" "$ACTUAL_OUTDIR" \
 
 (
     cd /outdir-base
-    {
-        echo "$GIT_ARCHIVE"
-        find "$ACTUAL_OUTDIR" -type f
-    } | xargs realpath --relative-base="$PWD" \
-      | xargs sha256sum \
+    find "$GIT_ARCHIVE" "$ACTUAL_OUTDIR" -type f -execdir bash -c 'basename "$1" | xargs sha256sum' _ {} \; \
```

Comment:
How does the `_` variable work here?

The Bash [doc](https://www.gnu.org/software/bash/manual/bash.html#Bash-Variables) did not help much.

Reply:
LOL I just tried for 30 min to figure out :) I am not bad in Bash, but this I just can not understand.

Reply:
AFAIK, post shell startup, `_` stores the arguments of the last executed command. E.g. it stores `-G` after the `ls -G` command is executed.

So in the case above, it should store the pathname `/outdir-base`, i.e. the argument of the last command executed before this line (`cd /outdir-base`). 

Unless I am missing something here.

Reply:
I guess the _ feeds the bash command, so it results in $1 there, but not sure :)

Reply:
I would assume it works something like that, but honestly not sure at this point ¯\\\_(ツ)_/¯.

Reply:
Shellcheck told me to do it this way: https://github.com/koalaman/shellcheck/wiki/SC2156

Reply:
> Shellcheck told me to do it this way: https://github.com/koalaman/shellcheck/wiki/SC2156

TIL, thanks!

Reply:
This is really Bash voodoo, I wonder how many people in the world understand this. Please add a comment, e.g. _"The _ is a dummy string that becomes $0 in the bash command. {} becomes $1 in the bash command. Recommended by shellcheck, see https://github.com/koalaman/shellcheck/wiki/SC2156"_ so more people can understand the code.

Reply:
ACK on adding that comment (or anything similar). It would save many hours of trying to grok this.

Reply:
I've added a comment.

Reply:
I had still the ambition to understand this strange `_` "dummy" parameter and to find any documentation about that, but as Hebasto said, Bash manpage does not help on `_`. After some testing I think I found out now :)
Manpage of Bash 5.0.3 for option `-c`:

> If the -c option is present, then commands are read from the first non-option argument command_string.
> If there are arguments after the command_string,  the first argument is assigned to $0
> and any remaining arguments are assigned to the positional parameters.
> The assignment to $0 sets the name of the shell, which is used in warning and error messages.

Unfortunately, I do not understand what the last sentence means, can anybody explain? But the text before the last sentence can be proofed with this example:
```
bash -c 'echo $0 $1' param1 param2
param1 param2
```
So what we know from Bash scripts, that $0 is the name of the script, is not the case in bash command using `-c`. In the latter case, $0 is the first parameter, what we know from Bash scripts as $1.
So I think the `_` is only used to fill $0, so the numbering of parameters using bash command -c is the same as in bash script (first really used parameter $1 and so on).
It is not really needed.
You can just replace $1 with $0 in the bash command, then you do not need the `_` dummy parameter.
So you can write shorter and simpler this:
`find "$GIT_ARCHIVE" "$ACTUAL_OUTDIR" -type f -execdir bash -c 'basename "$0" | xargs sha256sum' {} \; \`
Shellcheck is also happy with this.
I would suggest this, as it is simpler and shorter and less confusing. There is only need to explain that with `-c` what one normally knows as $1 becomes $0 there.
A comment to explain the parameter numbering shift is needed anyway I think, be it using the `_` dummy parameter or not. So one can comment like this:

`Unlike in Bash scripts, the first parameter to bash -c <command> is assigned to $0, not $1`

I updated the shellcheck [wiki page](https://github.com/koalaman/shellcheck/wiki/SC2156) to remove the confusing dummy parameter usage.

Reply:
I had still the ambition to understand this strange `_` "dummy" parameter and to find any documentation about that, but as Hebasto said, Bash manpage does not help on `_`. After some testing I think I found out now :)
Manpage of Bash 5.0.3 for option `-c`:

> If the -c option is present, then commands are read from the first non-option argument command_string.
> If there are arguments after the command_string,  the first argument is assigned to $0
> and any remaining arguments are assigned to the positional parameters.
> The assignment to $0 sets the name of the shell, which is used in warning and error messages.

Unfortunately, I do not understand what the last sentence means, can anybody explain? But the text before the last sentence can be proofed with this example:
```
bash -c 'echo $0 $1' param1 param2
param1 param2
```
So what we know from Bash scripts, that $0 is the name of the script, is not the case in bash command using `-c`. In the latter case, $0 is the first parameter, what we know from Bash scripts as $1.
So I think the `_` is only used to fill $0, so the numbering of parameters using bash command -c is the same as in bash script (first really used parameter $1 and so on).
It is not really needed.
You can just replace $1 with $0 in the bash command, then you do not need the `_` dummy parameter.
So you can write shorter and simpler this:
`find "$GIT_ARCHIVE" "$ACTUAL_OUTDIR" -type f -execdir bash -c 'basename "$0" | xargs sha256sum' {} \; \`
Shellcheck is also happy with this.
I would suggest this, as it is simpler and shorter and less confusing. There is only need to explain that with `-c` what one normally knows as $1 becomes $0 there.
A comment to explain the parameter numbering shift is needed anyway I think, be it using the `_` dummy parameter or not. So one can comment like this:

`Unlike in Bash scripts, the first parameter to bash -c <command> is assigned to $0, not $1`

I updated the shellcheck [wiki page](https://github.com/koalaman/shellcheck/wiki/SC2156) to remove the confusing dummy parameter usage.

Reply:
I think it's generally discouraged to used `$0` as an actual parameter. Bash's parameter expansion (through `$*` and `$@`) always starts with 1 because 0 is always the filename. If you consider a general case, people may want to use `$*` or `$@` in their commands, in which case there would be an off by one error if this was not mentioned. We should probably have this discussion in the shellcheck repo rather than here.

I'm going to leave this as is since I'm not confident the using `$0` is a good idea.

Reply:
OK with dummy parameter then, sounds reasonable, but I would recommend to not name it "_" but "dollarzero" or something, like I have written, to avoid the confusion seen here.

Reply:
Just tested and can confirm:
```
bash -c 'echo $*' param1 param2
param2
```

Reply:
So i think the confusion for non-Bash-Gurus, while technically being correct, can be boiled down to unluckily, because mistakable, dummy parameter name and insufficient documentation of the reasonability to use it An appropriate update of the shellcheck wiki page would be great.

Reply:
> Manpage of Bash 5.0.3 for option `-c`:
>
> > The assignment to $0 sets the name of the shell, which is used in warning and error messages.

I think what they mean is:
"$0 in the shell command is used as the name of the shell, which is used in warning and error messages."

Examples:
```
$ provoke_error
bash: provoke_error: command not found
# -> "bash" is the name of the shell in the error message

$ bash -c 'provoke_error' bash
bash: provoke_error: command not found
# -> "bash" is the name of the shell in the error message

$ bash -c 'provoke_error' mySuperShell
mySuperShell: provoke_error: command not found
# -> "mySuperShell" is the name of the shell in the error message

$ bash -c 'provoke_error' $SHELL
/bin/bash: provoke_error: command not found
# -> "/bin/bash" is the name of the shell in the error message,
#                       which was taken from the command invoking shell's environment variable $SHELL
```

Reply:
I have updated https://github.com/koalaman/shellcheck/wiki/SC2156

Reply:
I've changed this to use `bash` instead of `_`.